src/go/configgenerator/filtergen/cors.go

```diff
@@ -15,42 +15,135 @@
 package filtergen
 
 import (
+	"fmt"
+	"strconv"
+	"time"
+
 	"github.com/GoogleCloudPlatform/esp-v2/src/go/options"
-	"github.com/GoogleCloudPlatform/esp-v2/src/go/util/httppattern"
 	corspb "github.com/envoyproxy/go-control-plane/envoy/extensions/filters/http/cors/v3"
+	matcherpb "github.com/envoyproxy/go-control-plane/envoy/type/matcher/v3"
 	"github.com/golang/glog"
 	servicepb "google.golang.org/genproto/googleapis/api/serviceconfig"
 	"google.golang.org/protobuf/proto"
+	"google.golang.org/protobuf/types/known/wrapperspb"
 )
 
 const (
 	// CORSFilterName is the Envoy filter name for debug logging.
 	CORSFilterName = "envoy.filters.http.cors"
 )
 
-type CORSGenerator struct{}
+type CORSBasicGenerator struct {
+	// AllowOrigin is the name of the origin to allow through.
+	AllowOrigin string
+	Options     *CorsOptions
+
+	NoopFilterGenerator
+}
+
+type CORSRegexGenerator struct {
+	// AllowOriginRegex is the regex of the origins to allow through.
+	AllowOriginRegex string
+	Options          *CorsOptions
+
+	NoopFilterGenerator
+}
+
+type CorsOptions struct {
+	MaxAge           time.Duration
+	AllowMethods     string
+	AllowHeaders     string
+	ExposeHeaders    string
+	AllowCredentials bool
+}
 
 // NewCORSFilterGensFromOPConfig creates a CORSGenerator from
 // OP service config + descriptor + ESPv2 options. It is a FilterGeneratorOPFactory.
 func NewCORSFilterGensFromOPConfig(serviceConfig *servicepb.Service, opts options.ConfigGeneratorOptions) ([]FilterGenerator, error) {
-	if opts.CorsPreset != "basic" && opts.CorsPreset != "cors_with_regex" {
+	corsOptions := &CorsOptions{
+		MaxAge:           opts.CorsMaxAge,
+		AllowMethods:     opts.CorsAllowMethods,
+		AllowHeaders:     opts.CorsAllowHeaders,
+		ExposeHeaders:    opts.CorsExposeHeaders,
+		AllowCredentials: opts.CorsAllowCredentials,
+	}
+
+	switch opts.CorsPreset {
+	case "":
 		glog.Infof("Not adding CORS filter gen because the feature is disabled by option, option is currently %q", opts.CorsPreset)
 		return nil, nil
+
+	case "basic":
+		return []FilterGenerator{
+			&CORSBasicGenerator{
+				AllowOrigin: opts.CorsAllowOrigin,
+				Options:     corsOptions,
+			},
+		}, nil
+
+	case "cors_with_regex":
+		return []FilterGenerator{
+			&CORSRegexGenerator{
+				AllowOriginRegex: opts.CorsAllowOriginRegex,
+				Options:          corsOptions,
+			},
+		}, nil
+
+	default:
+		return nil, fmt.Errorf(`cors_preset must be either "basic" or "cors_with_regex"`)
 	}
+}
+
+func (g *CORSBasicGenerator) FilterName() string {
+	return CORSFilterName
+}
+
+func (g *CORSBasicGenerator) GenFilterConfig() (proto.Message, error) {
+	return &corspb.Cors{}, nil
+}
 
-	return []FilterGenerator{
-		&CORSGenerator{},
-	}, nil
+func (g *CORSBasicGenerator) GenPerHostConfig(vHostName string) (proto.Message, error) {
```

Comment:
2 cents: the readability could be better if put basic and regex in one class, and make `GenPerHostConfig` depends on opts.CorsPreset

Reply:
Done. I'm was trying to fix this problem, but I guess it's tech debt we can leave as-is.

```golang
type CORSGenerator struct {
	Preset string
	// AllowOrigin should only be set if preset=basic
	AllowOrigin string
	// AllowOriginRegex should only be set if preset=cors_with_regex
	AllowOriginRegex string
}
```